### PR TITLE
Add example notebook and README for real-time progress tracking with …

### PR DESCRIPTION
**Pull request by kapil27**
#### Summary
Adds an example notebook demonstrating how to monitor real-time training progress using Kubeflow Trainer v2 on Red Hat OpenShift AI.

#### What's included

1. Example notebook: Shows how to view and interpret training progress metrics
2. README: Setup instructions and CLI commands for progress monitoring

#### Key features demonstrated

1. Progress tracking is enabled by default with TransformersTrainer
2. Polling TrainJob annotations for real-time metrics
3. Interpreting progress data: steps, epochs, loss, throughput, ETA
4. No code changes required in training function

#### Example output
Progress: 75% | Step: 150/200 | Epoch: 1/2  Loss: 0.023 | Throughput: 58.2 samples/sec | ETA: 45s
#### Tested on

1. OpenShift AI cluster with Kubeflow Trainer v2
2. SDK version: v0.2.1+rhai0

### CODE REVIEW

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+    "\n",
+    "## Overview\n",
+    "\n",
+    "In this example, we fine-tune the **Qwen 2.5 1.5B Instruct** model on the **Stanford Alpaca** instruction-following dataset. The training runs on 2 GPU nodes with automatic progress tracking enabled, allowing you to monitor training metrics in real-time from the OpenShift AI Dashboard.\n",
+    "\n",
+    "### What You'll Learn\n",
+    "\n",
+    "| Feature | Description |\n",
+    "|---------|-------------|\n",
+    "| **Automatic Progress Tracking** | TransformersTrainer auto-injects a `KubeflowProgressCallback` that exposes training metrics via HTTP |\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: nit: this describes the implementation not what the users will learn.

```suggestion
    "| **Automatic Progress Tracking** | How to set up a training job with automatic progress tracking using TransformersTrainer. |\n",
```

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 2]
```diff
@@ -0,0 +1,935 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e31b9158",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Authentication configuration\n",
+    "api_server = os.getenv(\"OPENSHIFT_API_URL\")\n",
+    "token = os.getenv(\"NOTEBOOK_USER_TOKEN\")\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: suggestion: can we provide a cell for users to populate these values with a md cell telling them to set these values?

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 3]
```diff
@@ -0,0 +1,935 @@
...
+    "### Prerequisites\n",
+    "\n",
+    "Before running this notebook, ensure you have:\n",
+    "\n",
+    "1. **OpenShift AI Cluster** with Kubeflow Trainer v2 enabled\n",
+    "2. **Workbench** running Python 3.12+ with GPU access\n",
+    "3. **Environment Variables:**\n",
+    "   - `OPENSHIFT_API_URL` - Your OpenShift API server URL\n",
+    "   - `NOTEBOOK_USER_TOKEN` - Authentication token for API access\n",
+    "4. **Shared PVC** named `shared` mounted at `/opt/app-root/src/shared` in the workbench"
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: nit: can we provide a suggested size, and a link to instructions on how to create the PVC? It needs to be WRX right? Does the user need to specify any particular setting to make ti WRX?

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 4]
```diff
@@ -0,0 +1,935 @@
...
+    "configuration.host = api_server\n",
+    "configuration.verify_ssl = False  # Set to True if using trusted certificates\n",
+    "configuration.api_key = {\"authorization\": f\"Bearer {token}\"}\n",
+    "\n",
+    "# Mount paths explained:\n",
+    "# - Notebook sees PVC at /opt/app-root/src/shared (via workbench storage mount)\n",
+    "# - Training pods: SDK mounts PVC at /mnt/kubeflow-checkpoints (via pvc:// URI)\n",
+    "# - Using SDK's automatic mount avoids manual PodTemplateOverrides\n",
+    "NOTEBOOK_SHARED_PATH = \"/opt/app-root/src/shared\"  # Where notebook sees the shared PVC\n",
+    "SDK_MOUNT_PATH = \"/mnt/kubeflow-checkpoints\"  # Where SDK mounts PVC in training pods\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: question: how would the user know what these values are?

Is the notebook shared path what they configured when they created the pvc? If so, let's say that so they know.

Where does the sdk mount path come from?

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 5]
```diff
@@ -0,0 +1,935 @@
...
+   "metadata": {},
+   "source": [
+    "## Download Model and Dataset to Shared PVC\n",
+    "\n",
+    "Before submitting the training job, we pre-download the model and dataset to the shared PVC. This ensures:\n",
+    "- **Offline Training:** Training pods don't need internet access during training\n",
+    "- **Faster Startup:** No download delays when training pods start\n",
+    "- **Consistency:** All nodes use the same model weights and data\n",
+    "\n",
+    "### About the Dataset\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: suggestion: can we move this to be next to the "Model Details" bit so it doesn't interrupt the flow as much?

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 6]
```diff
@@ -0,0 +1,935 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "d6db38f5",
+   "metadata": {},
+   "source": [
+    "## Define the Training Function\n",
+    "\n",
+    "The training function runs inside each training pod as a distributed PyTorch process. TransformersTrainer serializes this function and executes it via `torchrun` on each node.\n",
+    "\n",
+    "### How Progress Tracking Works\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: suggestion: I suggest we move the explanation of how progress tracking works until the end. It's useful info, but it interrupts the flow of the example.

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 7]
```diff
@@ -0,0 +1,935 @@
...
+    "    - Injects KubeflowProgressCallback for real-time metrics\n",
+    "    - Injects checkpoint instrumentation when `TransformersTrainer.output_dir` is set\n",
+    "    - Auto-resumes from the latest checkpoint when available\n",
+    "    \"\"\"\n",
+    "    import os\n",
+    "\n",
+    "    # IMPORTANT: Set offline mode BEFORE importing transformers/huggingface_hub\n",
+    "    # This prevents the newer huggingface_hub from validating local paths as repo IDs\n",
+    "    os.environ[\"HF_HUB_OFFLINE\"] = \"1\"\n",
+    "    os.environ[\"TRANSFORMERS_OFFLINE\"] = \"1\"\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: nit: what about moving these to `TransformersTrainer` `env` argument? It removes the noise from the train func making it easier to follow.

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 8]
```diff
@@ -0,0 +1,935 @@
...
+    "trainer = TransformersTrainer(\n",
+    "    func=train_func,\n",
+    "    num_nodes=2,\n",
+    "    resources_per_node={\n",
+    "        \"nvidia.com/gpu\": 1,\n",
+    "        \"cpu\": \"4\",\n",
+    "        \"memory\": \"16Gi\",\n",
+    "    },\n",
+    "    enable_progression_tracking=True,\n",
+    "    metrics_poll_interval_seconds=30,\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: nit: is this the default value? If so I suggest we don't include it - keep the noise at a minimum. :)

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 9]
```diff
@@ -0,0 +1,935 @@
...
+    "        print(f\"     - {step.name}: {step.status}\")\n",
+    "    print()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d132f69b",
+   "metadata": {},
+   "source": [
+    "## Check Training Progress\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: suggestion: I think we should *not* include this section - the annotation is an internal implementation detail that's going to go away when we use the upstream implementation. We don't want people poking around and using it.

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 10]
```diff
@@ -0,0 +1,935 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1eccd277",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Delete the training job\n",
+    "client.delete_job(name=job_name)\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: question: should we suggest they delete the model from the PVC too?

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 11]
```diff
@@ -0,0 +1,935 @@
...
+    "    )\n",
+    "\n",
+    "    # NOTE: Do NOT set output_dir here - the SDK injects it automatically\n",
+    "    # when using TransformersTrainer with output_dir=\"pvc://...\" parameter.\n",
+    "    # The SDK mounts PVC at /mnt/kubeflow-checkpoints and configures checkpointing.\n",
+    "\n",
+    "    # NOTE: save_strategy, save_steps, save_total_limit are injected by SDK\n",
+    "    # via PeriodicCheckpointConfig - don't set them here to avoid conflicts\n",
+    "    training_args = TrainingArguments(\n",
+    "        output_dir=\"/mnt/kubeflow-checkpoints\",  # SDK overrides with pvc:// subpath\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: question: does this need setting? Doesn't the `TransformersTrainer` set this for us?

@efazal can you clarify?

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb [hunk 12]
```diff
@@ -0,0 +1,935 @@
...
+    "\n",
+    "In this example, we fine-tune the **Qwen 2.5 1.5B Instruct** model on the **Stanford Alpaca** instruction-following dataset. The training runs on 2 GPU nodes with automatic progress tracking enabled, allowing you to monitor training metrics in real-time from the OpenShift AI Dashboard.\n",
+    "\n",
+    "### What You'll Learn\n",
+    "\n",
+    "| Feature | Description |\n",
+    "|---------|-------------|\n",
+    "| **Automatic Progress Tracking** | TransformersTrainer auto-injects a `KubeflowProgressCallback` that exposes training metrics via HTTP |\n",
+    "| **Real-Time Metrics** | View current step, epoch, loss, and estimated time remaining in the OpenShift AI Dashboard |\n",
+    "| **PVC-Based Checkpointing** | Save model checkpoints to a shared PersistentVolumeClaim for durability and resume capability |\n",
```
Comment (robert-bell) [inline examples/trainer/progress-tracking/progress-tracking-example.ipynb]: question: should checkpointing be included in this example? I thought we were having a separate example for this?

My opinion is the smaller the example, the easier it is for a user to use and learn from.

**Review by robert-bell — CHANGES_REQUESTED** [review]
Overall lgtm. A few suggestions, most of which are mainly stylistic and minor but I think are worth addressing.

Thanks for putting this together @kapil27.

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 2 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: Thanks.. updated the same.. allowing user to add the openshift logging details

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 3 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: Agree.. Updated the suggested size and also redirected to readme.md for pvc setup instructions

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 4 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: added explanation and instructions to set pvc name if their PVC name is not shared.

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 5 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: Agree, updated the same.

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 6 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: make sense.. it will also allow user to have a recap of what they have performed at the end of the example.
updated the same.

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 9 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: agree, instead I have redirected them to view training progress from RHOAI dashboard

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 10 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: agree, updated the changes

File: examples/trainer/progress-tracking/progress-tracking-example.ipynb (hunk 12 above)
Comment (kapil27) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: since both training progression and JIT checkpointing will be present by default, making the information from the two concepts can easily get intertwined . I have made sure to explain only about progression tracking as per this example,.

Checkpointing is still present, but we keep the details minimal here. A separate checkpointing-focused example will cover how checkpointing works, when it triggers, and how to validate/resume from checkpoints in a way that’s clearer and more actionable.

wdyt?

Comment (robert-bell) [reply examples/trainer/progress-tracking/progress-tracking-example.ipynb]: sounds good to me. I'm happy with the current level of detail.

**Review by robert-bell — APPROVED** [review]
lgtm. G = great. Thanks @kapil27 for pushing this forward. This is really nice, clear work.
